### PR TITLE
feat(Core/Script): Hooks for custom arena teams/types

### PR DESCRIPTION
**Pull request by peti446**
## CHANGES PROPOSED:
-  Added script hooks for arena to create new arena team types. Useful for 1v1 arena mod to work over separate mode.

## TESTS PERFORMED:
- Testing all arena modes with and without the 1v1 arena module

## HOW TO TEST THE CHANGES:

1. Test different arenas and confirm they are working as intended

## Target branch(es):
- [x] Master

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by FrancescoBorzi** [comment]
I'd say making sure that the current Arenas (without any module) still work fine would be enough for accepting this PR,  just to be safe. No need to test the custom module as you already did it ;)

File: src/server/game/Battlegrounds/ArenaTeamMgr.cpp [hunk 1]
```diff
@@ -56,6 +56,25 @@ ArenaTeam* ArenaTeamMgr::GetArenaTeamByName(const std::string& arenaTeamName) co
...
+    for (ArenaTeamContainer::const_iterator itr = ArenaTeamStore.begin(); itr != ArenaTeamStore.end(); ++itr)
+    {
+        if (itr->second->GetType() != type)
+        {
+            continue;
+        }
+        std::string teamName = itr->second->GetName();
+        std::transform(teamName.begin(), teamName.end(), teamName.begin(), ::toupper);
+        if (search == teamName)
+            return itr->second;
```
Comment (Helias) [inline src/server/game/Battlegrounds/ArenaTeamMgr.cpp:73]: add curly brackets

File: src/server/game/Battlegrounds/ArenaTeamMgr.cpp [hunk 2]
```diff
@@ -65,6 +84,16 @@ ArenaTeam* ArenaTeamMgr::GetArenaTeamByCaptain(uint64 guid) const
     return nullptr;
 }
 
+
+ArenaTeam* ArenaTeamMgr::GetArenaTeamByCaptain(uint64 guid, const uint32 type) const
+{
+    for (ArenaTeamContainer::const_iterator itr = ArenaTeamStore.begin(); itr != ArenaTeamStore.end(); ++itr)
+        if (itr->second->GetCaptain() == guid && itr->second->GetType() == type)
+            return itr->second;
```
Comment (Helias) [inline src/server/game/Battlegrounds/ArenaTeamMgr.cpp:92]: here too

**Review by Helias — APPROVED** [review]

**Comment by Helias** [comment]
I would add that if anyone are testing this and can test this PR with the modules mod-arena-tol-viron and mod-arena-tiger-peak would be good :P

**Review by Helias — APPROVED** [review]

File: src/server/game/Scripting/ScriptMgr.h [hunk 3]
```diff
@@ -1668,20 +1668,18 @@ class ScriptRegistry
...
+            TScript* const script = *it;
 
             script->checkValidity();
 
-                if (script->IsDatabaseBound())
-                {
-                    if (!_checkMemory(script))
-                    {
-                        return;
-                    }
```
Comment (FrancescoBorzi) [inline src/server/game/Scripting/ScriptMgr.h]: please keep the brackets

File: src/server/game/Battlegrounds/ArenaTeamMgr.cpp [hunk 4]
```diff
@@ -51,17 +51,56 @@ ArenaTeam* ArenaTeamMgr::GetArenaTeamByName(const std::string& arenaTeamName) co
...
+    }
+    return nullptr;
+}
+
+
+ArenaTeam* ArenaTeamMgr::GetArenaTeamByName(std::string const& arenaTeamName, const uint32 type) const
+{
+    std::string search = arenaTeamName;
+    std::transform(search.begin(), search.end(), search.begin(), ::toupper);
+    for (ArenaTeamContainer::const_iterator itr = ArenaTeamStore.begin(); itr != ArenaTeamStore.end(); ++itr)
```
Comment (FrancescoBorzi) [inline src/server/game/Battlegrounds/ArenaTeamMgr.cpp:66]: you could use `auto` instead of `ArenaTeamContainer::const_iterator`

**Review by FrancescoBorzi — APPROVED** [review]

**Comment by FrancescoBorzi** [comment]
> I'd say making sure that the current Arenas (without any module) still work fine would be enough for accepting this PR, just to be safe. No need to test the custom module as you already did it ;)

@azerothcore/testers can anyone please test this? just test blizzlike Arenas and make sure nothing is broken with this PR.

**Comment by nicoaravena** [comment]
Tested this PR and it works almost fine.

---

**Ranked 2v2.**
Queue
imagen
Battle
imagen
Results
imagen
imagen

---

**Skirmish 2v2:**
Queue call
imagen
Battle
imagen
Results
There is a case that if the character is an insta 80, the results table did not show after reconnect, but it is just in that case, I think that is not related.
imagen

**Comment by LukasVolgger** [comment]
Thx @nicoaravena for your tests!

**Comment by Helias** [comment]
@nicoaravena you did a great job!

**Review by Si1ker — APPROVED** [review]

**Pull request merged**